extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/MongoDBSourceBuilder.java

```diff
@@ -217,645 +88,404 @@ public static MongoDBSourceBuilder<Void>.StreamDatabase<Void, Void> streamDataba
-            FunctionEx<? super ChangeStreamDocument<? extends T>, U> mapFn,
-            FunctionEx<? super MongoClient, ? extends BsonTimestamp> startAtOperationTimeFn
-
+            @Nonnull SupplierEx<? extends MongoClient> clientSupplier
     ) {
-        return () -> {
-            MongoClient client = connectionSupplier.get();
-            MongoDatabase database = databaseFn.apply(client);
-            MongoCollection<? extends T> collection = collectionFn.apply(database);
-            ChangeStreamIterable<? extends T> changeStreamIterable = searchFn.apply(collection);
-            return new StreamContext<>(client, changeStreamIterable, mapFn, destroyFn, startAtOperationTimeFn);
-        };
+        return new Stream<>(name, clientSupplier);
     }
 
-    private static <T, U> SupplierEx<StreamContext<T, U>> contextFn(
-            SupplierEx<? extends MongoClient> connectionSupplier,
-            FunctionEx<? super MongoClient, ? extends MongoDatabase> databaseFn,
-            ConsumerEx<? super MongoClient> destroyFn,
-            FunctionEx<? super MongoDatabase, ? extends ChangeStreamIterable<? extends T>> searchFn,
-            FunctionEx<? super ChangeStreamDocument<? extends T>, U> mapFn,
-            FunctionEx<? super MongoClient, ? extends BsonTimestamp> startAtOperationTimeFn
-    ) {
-        return () -> {
-            MongoClient client = connectionSupplier.get();
-            MongoDatabase database = databaseFn.apply(client);
-            ChangeStreamIterable<? extends T> changeStreamIterable = searchFn.apply(database);
-            return new StreamContext<>(client, changeStreamIterable, mapFn, destroyFn, startAtOperationTimeFn);
-        };
-    }
+    private abstract static class Base<T> {
 
-    private static <T, U> SupplierEx<StreamContext<T, U>> contextFn(
-            SupplierEx<? extends MongoClient> connectionSupplier,
-            ConsumerEx<? super MongoClient> destroyFn,
-            FunctionEx<? super MongoClient, ? extends ChangeStreamIterable<? extends T>> searchFn,
-            FunctionEx<? super ChangeStreamDocument<? extends T>, U> mapFn,
-            FunctionEx<? super MongoClient, ? extends BsonTimestamp> startAtOperationTimeFn
-    ) {
-        return () -> {
-            MongoClient client = connectionSupplier.get();
-            ChangeStreamIterable<? extends T> changeStreamIterable = searchFn.apply(client);
-            return new StreamContext<>(client, changeStreamIterable, mapFn, destroyFn, startAtOperationTimeFn);
-        };
-    }
+        protected String name;
+        protected SupplierEx<? extends MongoClient> clientSupplier;
+        protected Class<T> mongoType;
 
-    private abstract class Base<T> {
+        protected String databaseName;
+        protected String collectionName;
+        protected final List<Bson> aggregates = new ArrayList<>();
 
         private Base() {
         }
```

Comment:
Because it's the same class - works like invoking private method. See https://stackoverflow.com/questions/16661595/why-can-you-not-inherit-from-a-class-whose-constructor-is-private :)